File: android/app/src/main/java/vn/team/freechat/adapter/ContactListAdapter.java
package vn.team.freechat.adapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import tvd12.com.ezyfoxserver.client.R;
import vn.team.freechat.model.ChatContactModel;

/**
 * Created by tavandung12 on 10/3/18.
 */

public class ContactListAdapter extends ArrayAdapter<ChatContactModel> {

    private final List<ChatContactModel> items;
    private final Map<Long, ChatContactModel> itemMap;
    private static final int item_layout_id = R.layout.component_contact_list_item;

    public ContactListAdapter(Context context) {
        this(context, new ArrayList<ChatContactModel>());
    }

    public ContactListAdapter(Context context, List<ChatContactModel> items) {
        super(context, item_layout_id, items);
        this.items = items;
        this.itemMap = new HashMap<>();
        this.addItemModel(ChatContactModel.systemModel());
    }

    public ChatContactModel getItemModel(int position) {
        return items.get(position);
    }

    public void addItemModel(ChatContactModel item) {
        if(!itemMap.containsKey(item.getId())) {
            this.items.add(item);
            this.itemMap.put(item.getId(), item);
        }
    }

    public void addItemModels(Collection<ChatContactModel> items) {
        for(ChatContactModel item : items)
            addItemModel(item);
    }

    public void setItemModels(Collection<ChatContactModel> items) {
        this.items.clear();
        this.itemMap.clear();
        this.addItemModels(items);
    }

    @NonNull
    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        LayoutInflater inflater = LayoutInflater.from(getContext());
        View view = inflater.inflate(item_layout_id, parent, false);
        TextView usernameView = view.findViewById(R.id.username);
        TextView lastMessageView = view.findViewById(R.id.lastMessage);
        ChatContactModel model = items.get(position);
        usernameView.setText(model.getContactUser());
        lastMessageView.setText(model.getLastMessage());
        return view;
    }
}

File: kotlin/app/src/main/java/vn/team/freechat_kotlin/adapter/ContactListAdapter.kt
package vn.team.freechat_kotlin.adapter


import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView
import vn.team.freechat_kotlin.R
import vn.team.freechat_kotlin.model.ContactListItemModel

/**
 * Created by tavandung12 on 10/3/18.
 */

class ContactListAdapter(context: Context) : BaseAdapter() {

    private val contacts : MutableList<ContactListItemModel>
    private val contactByChannelId: MutableMap<Long, ContactListItemModel>
    private val inflater : LayoutInflater = LayoutInflater.from(context)

    init {
        this.contacts = ArrayList()
        this.contactByChannelId = HashMap()
        this.contacts.add(ContactListItemModel(0L, listOf("Bot")))
    }

    fun addItemModels(itemModels: Collection<ContactListItemModel>) {
        itemModels.forEach {
            if (!contactByChannelId.containsKey(it.channelId)) {
                contacts.add(it)
                contactByChannelId[it.channelId] = it
            }
        }
    }

    @SuppressLint("ViewHolder")
    override fun getView(position: Int, convertView: View?, parent: ViewGroup?) : View {
        val view = inflater.inflate(R.layout.component_contact_list_item, parent, false)
        val usernameView = view.findViewById<TextView>(R.id.username)
        val lastMessageView = view.findViewById<TextView>(R.id.lastMessage)
        val model = contacts[position]
        usernameView.text = model.getUsersString()
        lastMessageView.text = model.lastMessage
        return view
    }

    override fun getCount(): Int {
        return contacts.size
    }

    override fun getItem(position: Int): ContactListItemModel {
        return contacts[position]
    }

    override fun getItemId(position: Int): Long {
        val item = contacts[position]
        return item.channelId
    }
}
